Update/create stablecoin token sets.

thetroyharris opened the pull request:
## Thank you for contributing to Spellbook 🪄

### Contribution type
Please check the type of contribution this pull request is for:

- [x] New spell(s)
- [x] Adding to existing spell lineage
- [ ] Bug fix

**Note:** You can safely discard any section below which doesn't apply based on selection above

---

Adding stablecoins to existing chains and creating the necessary files for chains that do not currently have them.

dune-eng commented:
Workflow run id 9539966168 approved.

dune-eng commented:
Workflow run id 9539966255 approved.

dune-eng commented:
Workflow run id 9540079571 approved.

dune-eng commented:
Workflow run id 9540079664 approved.

dune-eng commented:
Workflow run id 9540106635 approved.

dune-eng commented:
Workflow run id 9540106704 approved.

dune-eng commented:
Workflow run id 9540155140 approved.

dune-eng commented:
Workflow run id 9540155251 approved.

thetroyharris commented:
Ready for review @jeff-dude @Hosuke

Hosuke reviewed (commented):
Do you want to add a roll-up spell `tokens.erc20_stablecoins` view?

dune-eng commented:
Workflow run id 9548623561 approved.

dune-eng commented:
Workflow run id 9548623719 approved.

dune-eng commented:
Workflow run id 9548927496 approved.

dune-eng commented:
Workflow run id 9548927852 approved.

dune-eng commented:
Workflow run id 9548947784 approved.

dune-eng commented:
Workflow run id 9548947962 approved.

dune-eng commented:
Workflow run id 9549083108 approved.

dune-eng commented:
Workflow run id 9549083328 approved.

thetroyharris commented:
> Do you want to add a roll-up spell `tokens.erc20_stablecoins` view?

Sure. Just created it! 
If this PR could get merged today that would be lovely as my PR depends on these spells existing! :D

Hosuke commented on tokens/models/tokens/tokens_erc20_stablecoins.sql:
```diff
@@ -0,0 +1,47 @@
+        schema = 'tokens'
+        , alias = 'erc20_stablecoins'
+        , materialized = 'table'
+        , post_hook='{{ expose_spells(\'["arbitrum","avalanche_c","base","blast","bnb","celo","ethereum","fantom","gnosis","optimism","polygon","scroll","zkevm","zksync","linea", "tron"]\',
+                        "sector",
+                        "tokens",
+                        \'["thetroyharris", "gentrexha", "dot2dotseurat", "msilb7", "lgingerich", "Henrystats"]\') }}'
+    )
```
```suggestion
        , post_hook='{{ expose_spells(blockchains = \'["arbitrum","avalanche_c","base","blast","bnb","celo","ethereum","fantom","gnosis","optimism","polygon","scroll","zkevm","zksync","linea", "tron"]\',
                        spell_type = "sector",
                        spell_name = "tokens",
                        contributors = \'["thetroyharris", "gentrexha", "dot2dotseurat", "msilb7", "lgingerich", "Henrystats"]\') }}'
    )
```

Hosuke reviewed (approved):
LGTM.✅
Thank you @thetroyharris 

[Sample tokens.erc20_stablecoins](https://dune.com/queries/3838039)

jeff-dude commented:
hi @thetroyharris 
can you explain the use case for these new static token spells? ideally we move away from this completely (we will on all erc20 once our source data is fixed / more complete).

do these tokens not exist already in `tokens.erc20`?

thetroyharris commented:
> hi @thetroyharris can you explain the use case for these new static token spells? ideally we move away from this completely (we will on all erc20 once our source data is fixed / more complete).
> 
> do these tokens not exist already in `tokens.erc20`?

Hey Jeff -- I am in the process of building some incremental stablecoin spells and need a stablecoin token set to filter larger tables e.g., `erc20_{{blockchain}}.evt_transfer`.

There is no good way to do this at the moment because even if I was to use the `tokens.erc20` table i would still need to filter it for stablecoins.

If the tokens table eventually contained a column for type of token i.e. rebase, stablecoin, etc. then I could use that, but currently that is not the case.

Creating these token sets was the fastest way to achieve a token set of stable coins.

dune-eng commented:
Workflow run id 9553492223 approved.

dune-eng commented:
Workflow run id 9553492448 approved.

dune-eng commented:
Workflow run id 9553550220 approved.

dune-eng commented:
Workflow run id 9553587268 approved.

dune-eng commented:
Workflow run id 9553587371 approved.

jeff-dude commented:
> Hey Jeff -- I am in the process of building some incremental stablecoin spells and need a stablecoin token set to filter larger tables e.g., `erc20_{{blockchain}}.evt_transfer`.
> 
> There is no good way to do this at the moment because even if I was to use the `tokens.erc20` table i would still need to filter it for stablecoins.
> 
> If the tokens table eventually contained a column for type of token i.e. rebase, stablecoin, etc. then I could use that, but currently that is not the case.
> 
> Creating these token sets was the fastest way to achieve a token set of stable coins.

okay, i see. rather than create in this format, can you create as views on top of `tokens.erc20` and filter down to the stablecoins? with either design, you need to manually maintain the list anyway (correct me if wrong here). it'll be cleaner for us to maintain in here as views on top

thetroyharris commented:
> okay, i see. rather than create in this format, can you create as views on top of `tokens.erc20` and filter down to the stablecoins? with either design, you need to manually maintain the list anyway (correct me if wrong here). it'll be cleaner for us to maintain in here as views on top

Sure. Would you like me to remove the existing Ethereum and Optimism stablecoins tables during this process?

I created the token lists to match what I saw already existed and just extend it to other chains.

jeff-dude commented:
> Sure. Would you like me to remove the existing Ethereum, Fantom, and Optimism stablecoins tables during this process?

that would be great, if feasible. we should double check the manual ones listed in existing chains are covered by `tokens.erc20`, so we don't drop any. we can add there as needed

dune-eng commented:
Workflow run id 9566312451 approved.

dune-eng commented:
Workflow run id 9566312754 approved.

dune-eng commented:
Workflow run id 9568957246 approved.

dune-eng commented:
Workflow run id 9568957371 approved.

dune-eng commented:
Workflow run id 9569018154 approved.

dune-eng commented:
Workflow run id 9569018478 approved.

dune-eng commented:
Workflow run id 9569106422 approved.

dune-eng commented:
Workflow run id 9569174894 approved.

dune-eng commented:
Workflow run id 9569175384 approved.

dune-eng commented:
Workflow run id 9569260988 approved.

dune-eng commented:
Workflow run id 9569261071 approved.

dune-eng commented:
Workflow run id 9569309033 approved.

dune-eng commented:
Workflow run id 9569309225 approved.

dune-eng commented:
Workflow run id 9569461879 approved.

dune-eng commented:
Workflow run id 9569462151 approved.

thetroyharris commented:
@jeff-dude 

I have changed this into a view. I removed the stablecoins tables from Fantom and Ethereum as they did not contain much more information than the erc20 table. Conversely, Optimism's stablecoin table contains much more information so that remains as is. Eventually, I would like my stablecoin view to contain that information but that task is not a priority for me at the moment.

This view is just laying the groundwork for the next PR.

thetroyharris commented:
Also, what do I have to do to run the `dbt-run` and `dbt-test` tests locally? Could you give me example of how to do it for this PR?

Hosuke commented:
> Also, what do I have to do to run the `dbt-run` and `dbt-test` tests locally? Could you give me example of how to do it for this PR?

Currently you can only run `dbt compile` locally to check syntax/compilation error.

thetroyharris commented:
> Currently you can only run `dbt compile` locally to check syntax/compilation error.

Ok ya I do that, it just doesn't catch the same errors as the tests here.

jeff-dude commented on tokens/models/tokens/arbitrum/tokens_arbitrum_erc20.sql:
```diff
@@ -79,4 +79,9 @@ FROM (VALUES
     , (0x323665443cef804a3b5206103304bd4872ea4253, 'USDV', 6)
     , (0x966570a84709d693463cdd69dcadb0121b2c9d26, 'taoUSD', 18)
     , (0x993614e1c8c9c5abe49462ce5702431978fd767f, 'S*ETH', 18)
+    , (0xaf88d065e77c8cc2239327c5edb3a432268e5831, 'USDC', 6) -- Native
```
i'm still a bit confused on this approach here. these stablecoins already exist in `tokens.erc20`:
```
select
  *
from
  tokens.erc20
where
  contract_address = 0xaf88d065e77c8cc2239327c5edb3a432268e5831
```

keep in mind that erc20 token metadata now combines an automated source + these static files to handle outlier, long tail tokens missing. i would validate all of these already exist in the table, then remove from this PR.

thetroyharris replied:
> keep in mind that erc20 token metadata now combines an automated source

Wow did I ever jump the gun on this. This hurts my soul.

jeff-dude commented on tokens/models/tokens/ethereum/tokens_ethereum_erc20_stablecoins.sql:
i'm not sure we should delete these, as they may have downstream dependencies or get used by a few users/dashboards on dune?

jeff-dude commented on tokens/models/tokens/linea/tokens_linea_erc20.sql:
similar to all other chains, i would check these already exist, but since this is a net new file it's worth a close look. we can remove the file altogether if they exist

jeff-dude commented on tokens/models/tokens/tokens_stablecoins.sql:
```diff
@@ -0,0 +1,211 @@
+    )
+}}
+
+SELECT contract_address, symbol, decimals, 'arbitrum' AS blockchain 
```
i'm not sure we need to take the hardcoded values approach again. the ideal scenario is to read from `tokens.erc20` and filter on specific tokens. maybe that's a jinja list at top of model of address/blockchain, then filter the main tokens table on the list via a for loop?

thetroyharris replied:
What is the best approach for me to create a spell that I can use to filter token transfers for contracts that are stablecoins?

The filtered table will then itself be a spell. Trying to do all of this in a query takes too long.

One way or another I have to manually keep a list of contract address that are stablecoins.

jeff-dude replied:
yes, agree to build an intermediate spell that contains only stablecoins. my ask is that we read from the source `tokens.erc20` and filter that down to stablecoins in this new spell, rather than start from scratch on a hardcoded list. that way we at least keep consistent on symbol/decimals.

you can get fancy with a jinja list variable and loop through that for the filter on the source table, essentially saying:
```
where
    (blockchain = blockchain and address = address)
    OR (blockchain = blockchain and address = address)
    ...
```

jeff-dude replied:
that would be preferred approach. if the variable isn't working, you can also do the `union all` approach and filter on a blockchain/addresses per union'd query

jeff-dude replied:
i would expect the files modified in this PR to be much fewer to get this built

jeff-dude commented:
closing due to inactivity. feel free to reopen and describe what needs to happen next, if necessary. thank you.
